### DISCUSSION TITLE
Elasticsearch support

### DISCUSSION THREAD

**Discussion by atduarte** [coollabsio/coolify#2087] [Feature Requests]
Elasticsearch is a very popular search engine which is used for many different use cases (full-text search, vector search, monitoring, log analysis, etc).

I would like to request that Coolify supports it out-of-the-box.

Considering the many use cases it supports I believe it would be important to:
1. Have support for data backup and restore.
2. Add a service for Kibana and maybe Logstash + Beats

**Comment by devdjdjdj** [comment]
I am working on adjusting [this](https://github.com/deviantony/docker-elk) to make it run on coolify.

Right now, it errors out:
```
[2024-Jun-24 20:38:44.900515] Starting deployment of Elk Stack to worker-1.
[2024-Jun-24 20:38:45.052000] Preparing container with helper image: ghcr.io/coollabsio/coolify-helper:latest.
[2024-Jun-24 20:38:46.386144] ----------------------------------------
[2024-Jun-24 20:38:46.387791] Importing deviantony/docker-elk:main (commit sha HEAD) to /artifacts/xg88ckw.
[2024-Jun-24 20:38:49.679333] Pulling & building required images.
[2024-Jun-24 20:39:02.442633] Oops something is not okay, are you okay? 😢

[2024-Jun-24 20:39:02.444778] Network xg88ckw_elk Creating
Network xg88ckw_elk Created
Volume "elasticsearch" Creating
Volume "elasticsearch" Created
Container elasticsearch-p40ossw Creating
time="2024-06-24T20:39:01Z" level=warning msg="mount of type `volume` should not define `bind` option"
Container elasticsearch-p40ossw Created
Container setup-p40ossw Creating
Container setup-p40ossw Created
Gracefully stopping... (press Ctrl+C again to force)
Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting "/data/coolify/applications/p40ossw/elasticsearch/config/elasticsearch.yml" to rootfs at "/usr/share/elasticsearch/config/elasticsearch.yml": mount /data/coolify/applications/p40ossw/elasticsearch/config/elasticsearch.yml:/usr/share/elasticsearch/config/elasticsearch.yml (via /proc/self/fd/6), flags: 0x5001: not a directory: unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type

```

**Comment by Trubador** [comment]
I second this request - I would really appreciate support for it as well as other databases like PostGreSQL.

**Comment by SvetZitrka** [comment]
Do you have a time estimate for the completion of testing and support deployment?

**Comment by Trubador** [comment]
I have managed to finally get elastic search and kibana running om coolify. Not really suited for production purposes but for now it works as intended. Backups still not implemented though

**Reply by SvetZitrka** [reply]
Some progress, I'm curious :)

**Comment by Trubador** [comment]
I can share some tips briefly on how I got Elastic Search and Kibana running:

- Set Elastic Search container to specific IP if docker DNS resolution between containers does not work as is my case currently, otherwise Kibana can't find the container
- Hardcode IPs in both Kibana and Elastic Search config files and setup as you normally would on a server by logging into the running container, you can possibly also set config using environment variables
- Don't use docker compose if neccessary as it is harder to debug
- Set a HTTPS://<generated/custom domain> domain on Coolify for the Kibana and Elastic Search application respectively 
- Make sure to update your hosting provider firewall rules to lock down the access but not that much you cant access Kibana or Elastic Search yourself.
- Dont set limit of how much memory Elastic and Kibana can use but make sure Elastic has 4.5 GB as a minimum and Kibana 1 GB
- Setup data volume for each on the Coolify VM with container paths: /usr/share/<kibana>/ and /usr/share/<elasticsearch>/

**Reply by SvetZitrka** [reply]
> I can share some tips briefly on how I got Elastic Search and Kibana running:
> 
> * Set Elastic Search container to specific IP if docker DNS resolution between containers does not work as is my case currently, otherwise Kibana can't find the container
> * Hardcode IPs in both Kibana and Elastic Search config files and setup as you normally would on a server by logging into the running container, you can possibly also set config using environment variables
> * Don't use docker compose if neccessary as it is harder to debug
> * Set a HTTPS://<generated/custom domain> domain on Coolify for the Kibana and Elastic Search application respectively
> * Make sure to update your hosting provider firewall rules to lock down the access but not that much you cant access Kibana or Elastic Search yourself.
> * Dont set limit of how much memory Elastic and Kibana can use but make sure Elastic has 4.5 GB as a minimum and Kibana 1 GB
> * Setup data volume for each on the Coolify VM with container paths: /usr/share// and /usr/share//

Is there any progress?

**Reply by Trubador** [reply]
I have the elasticsearch and kibana docker containers working with the above requirements. I dont have the time to delve into how to fix the limitations and create a production ready service template for coolify. You are very welcome to progress with that :)

**Comment by lars-sorensen** [comment]
I would really like to see this as well!

**Comment by vitezprchal** [comment]
Any update?

**Reply by Trubador** [reply]
I have moved on to using Algolia so I am not sure I will get around to implement this, but I have detailed all the necessary steps that need to be automated via a service template for coolify - so just build on top of that.

**Reply by Jordan-Hall** [reply]
`elasticsearch:
    image: "docker.elastic.co/elasticsearch/elasticsearch:8.12.2"
    environment:
      - node.name=${NODE_NAME:-elasticsearch}
      - cluster.name=${CLUSTER_NAME:-es-argilla-local}
      - discovery.type=${DISCOVERY_TYPE:-single-node}
      - "ES_JAVA_OPTS=${ES_JAVA_OPTS:-\"-Xms512m -Xmx512m\"}"
      - cluster.routing.allocation.disk.threshold_enabled=${CLUSTER_ROUTING_ALLOCATION_DISK_THRESHOLD_ENABLED:-false}
      - xpack.security.enabled=${XPACK_SECURITY_ENABLED:-false}
    ulimits:
      memlock:
        soft: -1
        hard: -1
    volumes:
      - "elasticsearch-data:/usr/share/elasticsearch/data/"
    healthcheck:
      test:
        - CMD-SHELL
        - "curl --silent --fail http://elasticsearch:9200"
      interval: 10s
      timeout: 10s
      retries: 5`
      
      Taken from an existing plugin maybe this would work

**Comment by tmallet** [comment]
I would love to have Elasticsearch in Coolify services!

**Comment by Slawikus** [comment]
I've added a PR with Elasticsearch service template https://github.com/coollabsio/coolify/pull/6300
